### ISSUE TITLE
Rule proposal: `prefer-temporal`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#1315]
The [Temporal](https://github.com/tc39/proposal-temporal) proposal is currently stage 3. It massively improves date and time handling in JavaScript.

There's a lot of potential for linting here.

We could block use of `Date` completely. Would probably have to be behind an option. But as far as I know, `Temporal` should cover everything `Date` currently handles.

## Fail

```js
Date.now()
```

## Pass

```js
Temporal.now.instant()
```

We could also potentially prevent use of popular date libraries like `moment`, `Luzon`, `date-fns`, etc.

More specific ideas for listing here wanted.

**Comment by fregante** [comment]
> We could also potentially prevent use of popular date libraries like `moment`, `Luzon`, `date-fns`, etc.

Honestly I'd love a `you-might-not-need-*`-style rules. There's plenty around lodash and jQuery

**Comment by fisker** [comment]
It seems `Temporal.Now.instant().epochMilliseconds` is the correct replacement for `Date.now()`.

**Comment by fregante** [comment]
On second thought, why? Why would `Temporal.Now.instant().epochMilliseconds` be better than `Date.now()`? It's just a number in both cases and there are no improvements (unlike the new `Number.*` methods).

I think such rule would make more sense for `Date` method _except_ `Date.now()`

**Comment by sindresorhus** [comment]
@fregante It has nanosecond accuracy instead of millisecond accuracy. It also has methods that makes it easy to compare, add/subtract, and format it. And consistency is a good argument too. And the name is just plain incorrect. `Date` is not actually dealing with dates. Java made a big mistake calling it "date" and a lot of languages copied it. "Instant" is the correct term.

You generally wouldn't actually use `Temporal.Now.instant().epochMilliseconds` as you rarely need the milliseconds, but you rather usually compare or format it in some way. So you generally use and store `Temporal.Now.instant()`.

I guess we could make it an option, but I personally can't think of any use-cases where I would prefer `Date.now()`.

**Comment by fregante** [comment]
> I personally can't think of any use-cases where I would prefer `Date.now()`.

I don't want to write code that looks like Java (ironically) (because it's verbose)


> So you generally use and store `Temporal.Now.instant()`.

This is still easier to read:


```js
const start = Date.now();
console.log(Date.now() - start);
```

compared to:


```js
const start = Temporal.now.instant();
console.log(start.compare(Temporal.now.instant());
```

> You generally wouldn't actually use `Temporal.Now.instant().epochMilliseconds` as you rarely need the milliseconds

Good point, but sometimes you still do, and `Date.now()` becomes `Temporal.Now.instant().epochMilliseconds`, which is mouthful and not scannable at all.


> It has nanosecond accuracy instead of millisecond accuracy

I think this is the only reason why it's better IMHO.

If the intention is _consistency over readability_, then sure, we agree, but I still think it's less readable.

**Comment by sindresorhus** [comment]
Luckily we don't have to decide this yet. We can make the option false at first and see how we feel about it after having used Temporal for a while.

**Comment by fisker** [comment]
How about just forbidding the use of `Date`, but allowing `Date.now()` behind an option?

**Comment by sindresorhus** [comment]
> How about just forbidding the use of Date, but allowing Date.now() behind an option?

👍
